pubtools/pulplib/_impl/client/client.py

```diff
@@ -181,10 +185,53 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def get_maintenance_report(self):
+        """Get the current maintenance mode status for this Pulp server.
+
+        Returns:
+            Future[:class:`~pubtools.pulplib.MaintenceReport`]
+                A future describes the maintenance status
+        """
+        report_ft = self._do_get_maintenance()
+
+        return f_map(
+            report_ft,
+            lambda data: MaintenanceReport._from_data(data)
+            if data
+            else MaintenanceReport(),
+        )
+
+    def set_maintenance(self, report):
+        """Set the current maintenance mode status for this Pulp server.
+
+        Args:
+            report:
+                An updated :class:`~pubtools.pulplib.MaintenanceReport` object that
+                will be used as the newest maintenance report.
+        Return:
+            Future[list[:class:`~pubtools.pulplib.Task`]]
+                A future which is resolved when maintenance mode has been updated successfully.
+
+                The future contains a task triggered and awaited during the publish
+                maintenance repository operation.
+        """
+        report_json = json.dumps(report._export_dict(), indent=4, sort_keys=True)
+        report_fileobj = StringIO(report_json)
+
+        repo = self.get_repository("redhat-maintenance").result()
+
+        # upload updated report to repository and publish
+        upload_ft = repo.upload_file(report_fileobj, "repos.json")
+
+        return f_flat_map(upload_ft, lambda _: repo.publish())
```

Comment:
Hmm... I don't think this is correct actually.  repo.publish() returns a *list of futures*, right?  Not a single future.  flat map isn't going to be able to flatten a list of futures into a single future.

So, what does the function actually return right now?  A Future[Task], Future[Future[Task]], Future[list[Future[Task]]] ... something else?

Reply:
Nope, I forgot how the API works.  repo.publish() returns a *future of list*.  So this should be OK.